src/shared/XLinkDevice.c

```diff
@@ -478,6 +480,8 @@ static linkId_t getNextAvailableLinkUniqueId()
     return INVALID_LINK_ID;
 }
 
+// BUGBUG not thread-safe, e.g. multiple threads from an app -- each thread creating
+// a connection to a different device
```

Comment:
Both `getNextAvailableLinkUniqueId` and `getNextAvailableLink` are quite broken indeed.
Went through and I think the best case is to combine them into one, remove the sem_init and call them before the PlatformConnect and "release/give back" the link if that fails afterwards.
Also use same mutex to protect other `availableXLinks` accesses, like `getLink` and `getLinkById`.

I'll test this out and open a PR - given that some below points might coincide with this, it might make sense opening it against your PR to combine modifications (TBD).

Reply:
@diablodale 
https://github.com/luxonis/XLink/pull/30
Opened separately. Intention to mostly just cover `XLinkConnect` threading issue

Reply:
Got it. I commented at PR #30